From df279a625c39dafff953fa3cdb2b60887fca3d30 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Fri, 7 Mar 2014 09:41:25 +0800
Subject: [PATCH 179/188] arm: am33xx: add reset resource detect.

AM33XX is a SoC between omap3 and omap4, and more close
to omap4. So, the specific reset resource bit should be
added to mach-related prm for watchdog.

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 arch/arm/mach-omap2/io.c               |    2 +
 arch/arm/mach-omap2/prm-regbits-33xx.h |    8 ++++
 arch/arm/mach-omap2/prm33xx.c          |   64 ++++++++++++++++++++++++++++++++
 arch/arm/mach-omap2/prm33xx.h          |    1 +
 4 files changed, 75 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/io.c b/arch/arm/mach-omap2/io.c
index 3eb9ae5..b1fa4e4 100644
--- a/arch/arm/mach-omap2/io.c
+++ b/arch/arm/mach-omap2/io.c
@@ -53,6 +53,7 @@
 #include "cminst44xx.h"
 #include "prm2xxx.h"
 #include "prm3xxx.h"
+#include "prm33xx.h"
 #include "prm44xx.h"
 
 /*
@@ -591,6 +592,7 @@ void __init am33xx_init_early(void)
 	omap2_set_globals_cm(AM33XX_L4_WK_IO_ADDRESS(AM33XX_PRCM_BASE), NULL);
 	omap3xxx_check_revision();
 	am33xx_check_features();
+	am33xx_prm_init();
 	am33xx_powerdomains_init();
 	am33xx_clockdomains_init();
 	am33xx_hwmod_init();
diff --git a/arch/arm/mach-omap2/prm-regbits-33xx.h b/arch/arm/mach-omap2/prm-regbits-33xx.h
index 84feece..ac5f14f 100644
--- a/arch/arm/mach-omap2/prm-regbits-33xx.h
+++ b/arch/arm/mach-omap2/prm-regbits-33xx.h
@@ -49,4 +49,12 @@
 #define AM33XX_RAM_MEM_ONSTATE_MASK			(0x3 << 30)
 #define AM33XX_RAM_MEM_RETSTATE_MASK			(1 << 27)
 #define AM33XX_RAM_MEM_STATEST_MASK			(0x3 << 21)
+
+
+#define AM33XX_GLOBAL_COLD_RST_SHIFT			0
+#define AM33XX_GLOBAL_WARM_SW_RST_SHIFT			1
+#define AM33XX_WDT1_RST_SHIFT				4
+#define AM33XX_EXTERNAL_WARM_RST_SHIFT			5
+#define AM33XX_ICEPICK_RST_SHIFT			9
+
 #endif
diff --git a/arch/arm/mach-omap2/prm33xx.c b/arch/arm/mach-omap2/prm33xx.c
index 7204407..d48269a 100644
--- a/arch/arm/mach-omap2/prm33xx.c
+++ b/arch/arm/mach-omap2/prm33xx.c
@@ -19,6 +19,7 @@
 #include <linux/err.h>
 #include <linux/io.h>
 
+#include "soc.h"
 #include "common.h"
 #include "powerdomain.h"
 #include "prm33xx.h"
@@ -343,3 +344,66 @@ struct pwrdm_ops am33xx_pwrdm_operations = {
 	.pwrdm_wait_transition		= am33xx_pwrdm_wait_transition,
 	.pwrdm_has_voltdm		= am33xx_check_vcvp,
 };
+
+/*
+ * am33xx_prm_reset_src_map - map from bits in the PRM_RSTST hardware
+ * register (which are specific to AM33xx SoCs) to reset source ID bit
+ * shifts (which is an OMAP SoC-independent enumeration)
+ */
+static struct prm_reset_src_map am33xx_prm_reset_src_map[] = {
+	{ AM33XX_GLOBAL_COLD_RST_SHIFT, OMAP_GLOBAL_COLD_RST_SRC_ID_SHIFT },
+	{ AM33XX_GLOBAL_WARM_SW_RST_SHIFT, OMAP_GLOBAL_WARM_RST_SRC_ID_SHIFT },
+	{ AM33XX_WDT1_RST_SHIFT, OMAP_SECU_WD_RST_SRC_ID_SHIFT },
+	{ AM33XX_EXTERNAL_WARM_RST_SHIFT, OMAP_EXTWARM_RST_SRC_ID_SHIFT },
+	{ AM33XX_ICEPICK_RST_SHIFT, OMAP_ICEPICK_RST_SRC_ID_SHIFT },
+	{ -1, -1 },
+};
+
+/**
+ * am33xx_prm_read_reset_sources - return the last SoC reset source
+ *
+ * Return a u32 representing the last reset sources of the SoC.  The
+ * returned reset source bits are standardized across OMAP SoCs.
+ */
+static u32 am33xx_prm_read_reset_sources(void)
+{
+	struct prm_reset_src_map *p;
+	u32 r = 0;
+	u32 v;
+ 
+	v = am33xx_prm_read_reg(AM33XX_PRM_DEVICE_MOD, AM33XX_PRM_RSTST_OFFSET);
+	p = am33xx_prm_reset_src_map;
+	while (p->reg_shift >= 0 && p->std_shift >= 0) {
+		if (v & (1 << p->reg_shift))
+			r |= 1 << p->std_shift;
+		p++;
+	}
+
+	return r;
+}
+
+/*
+ * XXX document
+ */
+static struct prm_ll_data am33xx_prm_ll_data = {
+	.read_reset_sources = &am33xx_prm_read_reset_sources,
+};
+
+int __init am33xx_prm_init(void)
+{
+	if (!soc_is_am33xx())
+		return 0;
+
+	return prm_register(&am33xx_prm_ll_data);
+}
+
+static void __exit am33xx_prm_exit(void)
+{
+	if (!soc_is_am33xx())
+		return;
+
+	/* Should never happen */
+	WARN(prm_unregister(&am33xx_prm_ll_data),
+	     "%s: prm_ll_data function pointer mismatch\n", __func__);
+}
+__exitcall(am33xx_prm_exit);
diff --git a/arch/arm/mach-omap2/prm33xx.h b/arch/arm/mach-omap2/prm33xx.h
index 9b9918d..b2f4593 100644
--- a/arch/arm/mach-omap2/prm33xx.h
+++ b/arch/arm/mach-omap2/prm33xx.h
@@ -127,5 +127,6 @@ extern int am33xx_prm_is_hardreset_asserted(u8 shift, s16 inst,
 extern int am33xx_prm_assert_hardreset(u8 shift, s16 inst, u16 rstctrl_offs);
 extern int am33xx_prm_deassert_hardreset(u8 shift, u8 st_shift, s16 inst,
 		u16 rstctrl_offs, u16 rstst_offs);
+extern int __init am33xx_prm_init(void);
 #endif /* ASSEMBLER */
 #endif
-- 
1.7.5.4

